File: 2020/03.js
const fs = require("fs");
const rows = fs
  .readFileSync("inputs/03.in", "utf8")
  .split("\n")
  .filter((x) => x.trim());

function countTrees(slopeX, slopeY) {
  let res, x, y;
  res = x = y = 0;
  while (y < rows.length - 1) {
    x += slopeX;
    y += slopeY;
    if (rows[y][x % rows[0].length] === "#") {
      res += 1;
    }
  }
  return res;
}

console.log("Part1:", countTrees(3, 1));
console.log(
  "Part2:",
  [
    countTrees(1, 1),
    countTrees(3, 1),
    countTrees(5, 1),
    countTrees(7, 1),
    countTrees(1, 2),
  ].reduce((total, val) => total * val)
);

File: 2020/03.py
from math import prod

with open('inputs/03.in', 'r') as f:
    rows = f.read().splitlines()

def count_trees(slope_x, slope_y):
    res = x = y = 0
    while y < len(rows)-1:
        x += slope_x
        y += slope_y
        if rows[y][x % len(rows[0])] == '#':
            res += 1
    return res

print('Part1:', count_trees(3, 1))
print('Part2:', prod([
    count_trees(1, 1),
    count_trees(3, 1),
    count_trees(5, 1),
    count_trees(7, 1),
    count_trees(1, 2),
]))
